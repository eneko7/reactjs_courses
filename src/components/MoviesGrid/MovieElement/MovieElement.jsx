import React, { Component } from 'react';
import propTypes from 'prop-types';
import ImagePoster from '../ImagePoster/ImagePoster';
import Genres from '../../Genres/Genres';
import styles from './MovieElement.scss';

class MovieElement extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpenMovie: false,
    };
    this.openMovieRef = React.createRef();
  }

  componentDidMount() {
    this.openMovieRef.current.addEventListener('click', this.openMovie);
  }

  componentWillUnmount() {
    this.openMovieRef.current.removeEventListener('click', this.openMovie);
  }

  openMovie = () => {
    this.setState(prevState => ({
      isOpenMovie: !prevState.isOpenMovie,
    }));
  };

  render() {
    const { data } = this.props;
    if (!data) {
      return null;
    }
    return (
      <div className={styles.movieElement} ref={this.openMovieRef}>
        <ImagePoster src={data.poster_path} alt={data.title} />
        <div className={styles.wrapperTitleYear}>
          <span className={styles.title}>{data.title.toUpperCase()}</span>
          <span className={styles.year}>
            {new Date(data.release_date).getFullYear()}
          </span>
        </div>
        <Genres data={data.genres} />
      </div>
    );
  }
}

MovieElement.defaultProps = {
  data: {},
};

MovieElement.propTypes = {
  data: propTypes.objectOf(
    propTypes.oneOfType([
      propTypes.string,
      propTypes.number,
      propTypes.array,
      propTypes.bool,
    ]),
  ),
};

export default MovieElement;
