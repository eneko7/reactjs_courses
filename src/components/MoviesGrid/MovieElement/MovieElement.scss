@import "../../../assets/constants/constants.scss";

.movie_element {
  width: 30%;
  background-color: #fff;
  margin-bottom: 5%;
  cursor: pointer;

  .wrapper_title_year {
    display: flex;
    justify-content: space-between;
    padding: 0 10px;
    margin-bottom: 20px;

    .title {
      font-family: $fontFamilyBold;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      text-transform: uppercase;
    }

    .year {
      font-family: $fontFamilyRegular;
      border: 1px solid;
      padding: 2px 10px;
    }
  }
}

.movie_element:focus {
  outline: none;
}

@media (max-width: $laptop) {
  .movie_element {
    width: 48%;
  }
}

@media (max-width: $tablet) {
  .movie_element {
    width: 49%;
  }
}

@media (max-width: $phone) {
  .movie_element {
    width: 100%;
  }
}
