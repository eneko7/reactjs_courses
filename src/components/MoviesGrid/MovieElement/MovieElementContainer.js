import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import MovieElement from './MovieElement';
import { fetchMovie } from '../../../modules/movie/movieActions';
import { fetchSearchMovies } from '../../../modules/movies/moviesActions';
import { getMovieState, getIsFetchingMovie, getIsFetchedMovie } from '../../../modules/movie/movieSelectors';

const mapStateToProps = (state) => {
  const movie = getMovieState(state);
  return {
    genres: movie.genres,
    movie,
    isFetchingMovie: getIsFetchingMovie(state),
    isFetchedMovie: getIsFetchedMovie(state),
  };
};

const mapDispatchToProps = {
  fetchMovie,
  fetchSearchMovies,

};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MovieElement));
