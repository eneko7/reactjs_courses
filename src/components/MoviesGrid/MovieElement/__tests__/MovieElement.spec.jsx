import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Enzyme, { configure, mount } from 'enzyme';
import MovieElement from '../MovieElement';

const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

configure({ adapter: new Adapter() });

const shallowRenderer = new ShallowRenderer();
const data = {
  id: 412059,
  title: 'Mobile Homes',
  tagline: '',
  vote_average: 0,
  vote_count: 1,
  release_date: '2018-04-04',
  poster_path: 'https://image.tmdb.org/t/p/w500/of72rtl3j0C9yJMlSJZHXnaByz8.jpg',
  overview: 'In forgotten towns along the American border, a young mother drifts from one motel to the next with her intoxicating boyfriend and her 8-year-old son. The makeshift family scrapes by, living one hustle at a time, until the discovery of a mobile home community offers an alternative life.',
  budget: 0,
  revenue: 0,
  genres: [
    'Drama',
  ],
  runtime: 105,
};

jest.mock('../../ImagePoster/ImagePoster', () => () => <div>test</div>);

describe('MovieElement', () => {
  it('renders correctly', () => {
    shallowRenderer.render(<MovieElement data={data} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
  it('renders without data correctly', () => {
    shallowRenderer.render(<MovieElement data={null} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('MovieElement -> Click on Movie card (open)', () => {
    const wrapper = mount((
      <MovieElement data={data} />
    ));
    const element = wrapper.find('.movie_element').simulate('click');
    expect(element).toMatchSnapshot();
  });
});
