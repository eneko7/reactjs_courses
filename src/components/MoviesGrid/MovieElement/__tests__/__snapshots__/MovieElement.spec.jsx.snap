// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieElement MovieElement -> Click on Movie card (open) 1`] = `ReactWrapper {}`;

exports[`MovieElement renders correctly 1`] = `
<div
  className="movie_element"
  onClick={[Function]}
  role="presentation"
>
  <Unknown
    alt="Mobile Homes"
    src="https://image.tmdb.org/t/p/w500/of72rtl3j0C9yJMlSJZHXnaByz8.jpg"
  />
  <div
    className="wrapper_title_year"
  >
    <span
      className="title"
    >
      Mobile Homes
    </span>
    <span
      className="year"
    >
      2018
    </span>
  </div>
  <Genres
    data={
      Array [
        "Drama",
      ]
    }
  />
</div>
`;

exports[`MovieElement renders without data correctly 1`] = `null`;
